plugins/colorbutton/plugin.js

```diff
@@ -370,6 +414,152 @@ CKEDITOR.plugins.add( 'colorbutton', {
 			}
 		}
 
+		function renderContentColors( options ) {
+			// This function is called on each dialog opening.
+			var contentColorsRow = options.contentColorsRow,
+				contentColorsLabel = options.contentColorsLabel,
+				cssAttribute = options.cssAttribute,
+				clickFn = options.clickFn,
+				type = options.type,
+				colorSpans = editor.editable().find( 'span[style*=' + cssAttribute + ']' ).toArray(),
+				colors = [],
+				colorOccurrences = {},
+				sortedColors = [],
+				colorsPerRow = config.colorButton_colorsPerRow || 6,
+				tilePosition = 1;
+
+			// It's more reliable and easier to delete exisiting color tiles and recreate them than to manipulate the old ones.
+			CKEDITOR.tools.array.forEach( contentColorsRow.find( 'td' ).toArray(), function( node ) {
+				node.remove();
+			}, 0, 1 );
+
+			if ( !colorSpans.length ) {
+				contentColorsLabel.hide();
+				return;
+			}
+
+			CKEDITOR.tools.array.forEach( colorSpans, function( span ) {
+				// colorSpans may contain spans with background color when they are not necessary, so here they are being filtered out.
+				var spanColor = span.getStyle( cssAttribute );
+
+				if ( spanColor ) {
+					colors.push( normalizeColor( spanColor ).toUpperCase() );
```

Comment:
The next `forEach` loop which calculates occurrences could be integrated somewhere here.
It would prevent of one unnecessary of iteration over the same colors twice. Once for spans and then once for read values.